### PR TITLE
init: fixes file descriptor accounting

### LINKED ISSUE #18911: Improve minimum file descriptor accounting and documentation
#16003 has been closed, and I've pulled out the unrelated p2p change. However I think it's worth someone following up and taking a look at our startup file descriptor accounting. 

As mentioned in #16003, if `bitcoind` is started somewhere that the maximum number of open file descriptors is limited, we can run into some nonsensical behaviour. Such as a "negative" number of maximum connections:

```bash
# using bash
ulimit -n 150
src/bitcoind
[init] Using at most -8 automatic connections (150 file descriptors available)
```

It's also impossible to shutdown `bitcoind` from this state as the `opencon` thread will never exit.

It would be good for someone to look through this code, document assumptions, and fix issues like the above.

You do not need to request permission to start working on this. You are encouraged to comment on the issue if you are planning to work on it. This will help other contributors monitor which issues are actively being addressed and is also an effective way to request assistance if and when you need it.

For guidance on contributing, please read [CONTRIBUTING.md](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md) before opening your pull request.

### PR DESCRIPTION
**Pull request by sr-gi**
The current logic for file descriptor accounting is pretty convoluted and hard to follow. This is partially caused by the lack of documentation plus non-intuitive variable naming (which was more intuitive when fewer things were accounted for, but
hasn't aged well). This has led to this accounting being error-prone and hard to maintain (as shown in the first commit of this PR).

Redefine some of the constants, plus document what are we accounting for so this can be extended more easily

Fixes #18911

### CODE REVIEW

File: src/init.cpp [hunk 1]
```diff
@@ -973,18 +973,16 @@ bool AppInitParameterInteraction(const ArgsManager& args)
...
+    // If we are using select instead of poll, our actuall limit may be even smaller
+#ifndef USE_POLL
+    nFD = std::min(FD_SETSIZE, nFD);
 #endif
-    // Trim requested connection counts, to fit into system limitations
-    // <int> in std::min<int>(...) to work around FreeBSD compilation issue described in #2695
-    nMaxConnections = std::max(std::min<int>(nMaxConnections, fd_max - nBind - MIN_CORE_FILEDESCRIPTORS - MAX_ADDNODE_CONNECTIONS - NUM_FDS_MESSAGE_CAPTURE), 0);
     if (nFD < MIN_CORE_FILEDESCRIPTORS)
         return InitError(_("Not enough file descriptors available."));
-    nMaxConnections = std::min(nFD - MIN_CORE_FILEDESCRIPTORS - MAX_ADDNODE_CONNECTIONS - NUM_FDS_MESSAGE_CAPTURE, nMaxConnections);
```
Comment (sr-gi) [inline src/init.cpp]: Notice how the latter is redundant once we limit `nFD = std::min(FD_SETSIZE, nFD);`. Moving things around, plus accounting for `nBind` (which is currently not being accounted for in master), we get:

```
nMaxConnections = std::max(std::min<int>(nMaxConnections, nFD - nBind - MIN_CORE_FILEDESCRIPTORS - MAX_ADDNODE_CONNECTIONS - NUM_FDS_MESSAGE_CAPTURE), 0);
nMaxConnections = std::min(nMaxConnections, nFD - nBind - MIN_CORE_FILEDESCRIPTORS - MAX_ADDNODE_CONNECTIONS - NUM_FDS_MESSAGE_CAPTURE,);
```

Comment (sr-gi) [inline src/init.cpp]: Notice how the latter `nMaxConnections` trim is redundant once we limit `nFD = std::min(FD_SETSIZE, nFD);`. Moving things around, plus accounting for `nBind` (which is currently not being accounted for in master), we get:

```
nMaxConnections = std::max(std::min<int>(nMaxConnections, nFD - nBind - MIN_CORE_FILEDESCRIPTORS - MAX_ADDNODE_CONNECTIONS - NUM_FDS_MESSAGE_CAPTURE), 0);
nMaxConnections = std::min(nMaxConnections, nFD - nBind - MIN_CORE_FILEDESCRIPTORS - MAX_ADDNODE_CONNECTIONS - NUM_FDS_MESSAGE_CAPTURE,);
```

**Comment by TheCharlatan** [comment]
Concept ACK

**Comment by sr-gi** [comment]
I think it is worth mentioning that the current approach is defining the minimum amount of file descriptions required without accounting for a single connection*, this means that if we are at the bare minimum, we will run but won't be able to connect to any nodes. This is consistent with the current logic in master, but I think it's not the way it should be.

I'm happy to add another commit addressing this, but I've rathered start approaching this sticking to the same assumptions as master.

* This actually accounts for manual connections, which may never happen, but not to any of our outbound. If we happen to not create any manuals we'd have 8 slots (`MAX_ADDNODE_CONNECTIONS`) for outbounds

**Comment by sr-gi** [comment]
@vasild you may be interested in this. I decided to fix it when seeing you extending the current logic in #29415

**Review by BrandonOdiwuor — COMMENTED** [review]
Concept ACK

File: src/init.cpp [hunk 2]
```diff
@@ -967,24 +968,29 @@ bool AppInitParameterInteraction(const ArgsManager& args)
...
+    nFD = RaiseFileDescriptorLimit(nMaxConnections + nMinRequiredFds);
+    // If we are using select instead of poll, our actuall limit may be even smaller
+#ifndef USE_POLL
+    nFD = std::min(FD_SETSIZE, nFD);
 #endif
+    if (nFD < MIN_CORE_FDS)
+        return InitError(strprintf(_("Not enough file descriptors available. %d available, %d required."), nFD, MIN_CORE_FDS));
+
     // Trim requested connection counts, to fit into system limitations
     // <int> in std::min<int>(...) to work around FreeBSD compilation issue described in #2695
```
[the remark this answers is not in the document]
Comment (sr-gi) [reply src/init.cpp]: Covered in [55f16f5](https://github.com/bitcoin/bitcoin/pull/30065/commits/55f16f56f4abeae1f5560cc616ea0b9b31a88073)

**Comment by vasild** [comment]
> The current logic for file descriptor accounting is pretty convoluted and hard to follow.

Concept ACK, I stumbled on this recently in https://github.com/bitcoin/bitcoin/pull/29415

File: src/init.cpp [hunk 3]
```diff
@@ -967,24 +968,28 @@ bool AppInitParameterInteraction(const ArgsManager& args)
...
-
-#ifdef USE_POLL
-    int fd_max = nFD;
-#else
-    int fd_max = FD_SETSIZE;
+    // Reserve enough FDs to account for the bare minimum, plus any manual connections, plus the bound interfaces
+    int nMinRequiredFds = MIN_CORE_FDS + MAX_ADDNODE_CONNECTIONS + nBind;
+
+    // Try raising the FD limit to what we need (nFD may be smaller than the requested amount if this fails)
+    nFD = RaiseFileDescriptorLimit(nMaxConnections + nMinRequiredFds);
```
Comment (vasild) [inline src/init.cpp:983]: snake_case for variables. I think `available` or `available_fds` is a good name:

```suggestion
    available_fds = RaiseFileDescriptorLimit(nMaxConnections + nMinRequiredFds);
```

**Review by vasild — COMMENTED** [review]
Approach ACK 55f16f56f4

Might as well squash the last commit 55f16f56f4 `init: Remove FreeBSD workaround to #2695` into some of the previous ones that touch that line.

File: src/init.cpp (hunk 3 above)
Comment (sr-gi) [reply src/init.cpp:983]: I'm happy to rename `nFD`, but I kept it because it is not locally defined, but defined in the namespace, so at least one other method is using it. If we don't mind a bigger diff I'm happy to change it

File: src/init.cpp [hunk 4]
```diff
@@ -967,24 +968,28 @@ bool AppInitParameterInteraction(const ArgsManager& args)
...
     nMaxConnections = std::max(nUserMaxConnections, 0);
-
-    nFD = RaiseFileDescriptorLimit(nMaxConnections + MIN_CORE_FILEDESCRIPTORS + MAX_ADDNODE_CONNECTIONS + nBind + NUM_FDS_MESSAGE_CAPTURE);
-
-#ifdef USE_POLL
-    int fd_max = nFD;
-#else
-    int fd_max = FD_SETSIZE;
+    // Reserve enough FDs to account for the bare minimum, plus any manual connections, plus the bound interfaces
+    int nMinRequiredFds = MIN_CORE_FDS + MAX_ADDNODE_CONNECTIONS + nBind;
```
[the remark this answers is not in the document]
Comment (sr-gi) [reply src/init.cpp:980]: I agree the if check should be updated, but I don't necessarily agree with splitting `MAX_ADDNODE_CONNECTIONS` from `min_required_fds`. 

There is the outstanding question of whether the minimum required fds to run Core should account for some of the connections. Currently, we are not accounting for any (given the check is performed against `MIN_CORE_FDS`), but I think we should be accounting for at least, outbound + manuals

File: src/init.cpp [hunk 5]
```diff
@@ -967,24 +968,28 @@ bool AppInitParameterInteraction(const ArgsManager& args)
...
 
-    // Make sure enough file descriptors are available
+    // Make sure enough file descriptors are available. We need to reserve enough FDs to account for the bare minimum,
+    // plus all manual connections and all bound interfaces. Any remainder will be available for connection sockets
+
+    // Number of bound interfaces (we have at least one)
     int nBind = std::max(nUserBind, size_t(1));
+    // Maximum number of connetions with other nodes, this accounts for all types of outbound and inbounds expect for manual
     nUserMaxConnections = args.GetIntArg("-maxconnections", DEFAULT_MAX_PEER_CONNECTIONS);
     nMaxConnections = std::max(nUserMaxConnections, 0);
```
[the remark this answers is not in the document]
Comment (sr-gi) [reply src/init.cpp:978]: I think that makes more sense

**Comment by sr-gi** [comment]
Thanks for the suggestions @vasild, I've partially covered them and commented on some of the pending ones

**Review by vasild — APPROVED** [review]
ACK fa42bf7ee89999b33d4af0600b0d8da1275e5d5e

This is already an improvement over `master`. Would be nice to resolve https://github.com/bitcoin/bitcoin/pull/30065#discussion_r1613422626 as well.

File: src/init.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (sr-gi) [reply src/init.cpp]: I've fixed that, plus increased the `min_required_fds` to also account for automatic outbounds in the last commit. I'm open to discuss about the latter, but I think it's something we should also account for

[the remark this answers is not in the document]
Comment (vasild) [reply src/init.cpp]: The newly added `int auto_outbounds = std::min(MAX_OUTBOUND_FULL_RELAY_CONNECTIONS + MAX_BLOCK_RELAY_ONLY_CONNECTIONS + MAX_FEELER_CONNECTIONS, nUserMaxConnections);` seems to be duplicating this logic, but not exactly:

https://github.com/bitcoin/bitcoin/blob/be100cf4c77a3e45750773689e0a396fda39d8a7/src/net.h#L1069-L1073

and I think this is getting complicated again. I would suggest to keep it simple - drop the last commit f569b06a53851f844219cd80cd33676470e3b776 `init: fix min required fds check and account for outbounds`

and only apply this change to 3982543007 `init: improves file descriptors accounting and docs`:

```diff
-    if (available_fds < MIN_CORE_FDS)
+    if (available_fds < min_required_fds)
```

[the remark this answers is not in the document]
Comment (sr-gi) [reply src/init.cpp]: It is duplicating part of that, given `CConnman::m_max_automatic_outbound` is private.

I agree it is adding more complexity, but not accounting for it in the minimum means that could be in a situation where no single outbound connection is accounted for, so the user is basically only connected to manuals.

I acknowledge that's fairly rare, but the two reasons why we are currently not facing this issue is, first, because the minimum is fairly low, so it would need a system with really limited resources to become an issue, and second, because If we were in that situation it would also need to be that case that the node is defining manual connections, otherwise the 8 reserved slots for manual would be available for automatic.

Despite that, this feels like bad design, so I'd argue accounting for automatic is something we should do (though not necessarily as done in the current approach)

[the remark this answers is not in the document]
Comment (sr-gi) [reply src/init.cpp]: This could be a simpler alternative, that accounts for the automatic within the minimum

```diff
// Automatic outbounds are already accounted for in nUserMaxConnections, but we want to make sure they count towards the minimum later on
int auto_outbounds = std::min(MAX_OUTBOUND_FULL_RELAY_CONNECTIONS + MAX_BLOCK_RELAY_ONLY_CONNECTIONS + MAX_FEELER_CONNECTIONS, nUserMaxConnections);
-    // Reserve enough FDs to account for the bare minimum, plus any manual connections, plus the bound interfaces
-    int min_required_fds = MIN_CORE_FDS + MAX_ADDNODE_CONNECTIONS + nBind;
+    nUserMaxConnections -= auto_outbounds;
+
+    // Reserve enough FDs to account for the bare minimum, plus any manual connections, automatic outbounds and the bound interfaces
+    int min_required_fds = MIN_CORE_FDS + MAX_ADDNODE_CONNECTIONS + auto_outbounds + nBind;

[...]

-    if (available_fds < min_required_fds + auto_outbounds)
-        return InitError(strprintf(_("Not enough file descriptors available. %d available, %d required."), available_fds, min_required_fds + auto_outbounds));
+    if (available_fds < min_required_fds)
+        return InitError(strprintf(_("Not enough file descriptors available. %d available, %d required."), available_fds, min_required_fds));
```

[the remark this answers is not in the document]
Comment (vasild) [reply src/init.cpp]: If the node is connected to 8 manually defined outbound peers and 0 inbounds and 0 automatic outbounds, then it is still operational? No need to prohibit startup in this case?

[the remark this answers is not in the document]
Comment (sr-gi) [reply src/init.cpp]: Is it though? If the user has not specified that he wants no automatic connections (`-maxconnections != 0`) I'd argue he'd expect them to happen.

In any case, I think this is such an edge case that it is not worth blocking the PR on it. If you have a strong opinion on this, I'm happy to drop it and squash the rest of the commit to one of the previous ones

[the remark this answers is not in the document]
Comment (vasild) [reply src/init.cpp]: I am somewhat ~0 on erroring in the case of 8manual/0in/0out, but my bigger anxiety with the latest commit is that it makes it confusing again (or too complicated in my judgement). I tried to code a simple way to account for the automatic outbounds but gave up. Also, most naturally, as the "minimum required" name suggests, the check should be `if (... < min_required)`. I find `if (... < min_required + something more)` confusing.

[the remark this answers is not in the document]
Comment (sr-gi) [reply src/init.cpp]: I agree with the "making it more confusing". Given this is an edge case, I'll remove the outbound count and we can account for it if someone has a better way of doing so, or if someone opens an issue about it.

**Comment by sr-gi** [comment]
> ACK [fa42bf7](https://github.com/bitcoin/bitcoin/commit/fa42bf7ee89999b33d4af0600b0d8da1275e5d5e)
> 
> This is already an improvement over `master`. Would be nice to resolve [#30065 (comment)](https://github.com/bitcoin/bitcoin/pull/30065#discussion_r1613422626) as well.

Addressed all comments now

File: src/init.cpp [hunk 6]
```diff
@@ -967,24 +968,28 @@ bool AppInitParameterInteraction(const ArgsManager& args)
...
+    if (nFD < MIN_CORE_FDS)
+        return InitError(strprintf(_("Not enough file descriptors available. %d available, %d required."), nFD, MIN_CORE_FDS));
+
     // Trim requested connection counts, to fit into system limitations
-    // <int> in std::min<int>(...) to work around FreeBSD compilation issue described in #2695
-    nMaxConnections = std::max(std::min<int>(nMaxConnections, fd_max - nBind - MIN_CORE_FILEDESCRIPTORS - MAX_ADDNODE_CONNECTIONS - NUM_FDS_MESSAGE_CAPTURE), 0);
-    if (nFD < MIN_CORE_FILEDESCRIPTORS)
-        return InitError(_("Not enough file descriptors available."));
-    nMaxConnections = std::min(nFD - MIN_CORE_FILEDESCRIPTORS - MAX_ADDNODE_CONNECTIONS - NUM_FDS_MESSAGE_CAPTURE, nMaxConnections);
+    nMaxConnections = std::max(std::min(nMaxConnections, nFD - nMinRequiredFds), 0);
```
[the remark this answers is not in the document]
Comment (sr-gi) [reply src/init.cpp]: I've rolled back this. After some of the refactoring, this looked like:

```
std::clamp(available_fds - min_required_fds, 0, nUserMaxConnections);
```

However, `available_fds` cannot be smaller than `min_required_fds`, and `nUserMaxConnections` cannot be negative. So calling `std::min` should be enough.

**Review by vasild — APPROVED** [review]
ACK a9e211a5302630635b825655cd984683794aa42c

Note: this is the last but one commit. The last commit 23f843d816 `init: fix min required fds check and account for outbounds` now contains a single dummy change of removing an empty line. Can/should be dropped altogether. You can `push -f` a9e211a5302630635b825655cd984683794aa42c into this branch.

Thanks!

File: src/init.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (vasild) [reply src/init.cpp]: Excellent! Some simplifications leading to further simplifications. :heart:

**Comment by sr-gi** [comment]
Nice catch @vasild, I've squashed that into one of the previous commits. It should be good now.

**Review by vasild — APPROVED** [review]
ACK fe76929d4d68a4ca6d27c58c1e243a64fdc70b2a

File: src/init.cpp [hunk 7]
```diff
@@ -967,24 +968,30 @@ bool AppInitParameterInteraction(const ArgsManager& args)
...
     }
 
-    // Make sure enough file descriptors are available
+    // Make sure enough file descriptors are available. We need to reserve enough FDs to account for the bare minimum,
+    // plus all manual connections and all bound interfaces. Any remainder will be available for connection sockets
+
+    // Number of bound interfaces (we have at least one)
     int nBind = std::max(nUserBind, size_t(1));
+    // Maximum number of connections with other nodes, this accounts for all types of outbounds and inbounds except for manual
     nUserMaxConnections = args.GetIntArg("-maxconnections", DEFAULT_MAX_PEER_CONNECTIONS);
```
Comment (TheCharlatan) [inline src/init.cpp:977]: I think this should be de-globalized and renamed like I've done here: https://github.com/TheCharlatan/bitcoin/commit/a2c048322c67436e02a3a9607400f65f10dd6564

The `avaialble_fds`  global variable there are also kind of silly, since it is only there to be logged at a later point in time, but I don't feel strongly if it should go or not either way.

**Review by sedited — COMMENTED** [review]
This looks good, but I think these few, simple things should be addressed.

**Comment by sr-gi** [comment]
Took (most of?) your suggestions @TheCharlatan

File: src/init.cpp [hunk 8]
```diff
@@ -142,11 +142,12 @@ static constexpr bool DEFAULT_STOPAFTERBLOCKIMPORT{false};
...
 // accessing block files don't count towards the fd_set size limit
 // anyway.
-#define MIN_CORE_FILEDESCRIPTORS 0
+#define MIN_LEVELDB_FDS 0
 #else
-#define MIN_CORE_FILEDESCRIPTORS 150
+#define MIN_LEVELDB_FDS 150
 #endif
 
+static const int MIN_CORE_FDS = MIN_LEVELDB_FDS + NUM_FDS_MESSAGE_CAPTURE;
```
[the remark this answers is not in the document]
Comment (TheCharlatan) [reply src/init.cpp]: Oh, that makes sense!

**Comment by sr-gi** [comment]
Rebased to make CI green. This was based on a pre-CMake commit

**Review by vasild — APPROVED** [review]
ACK d4c7c4009da14c84576d43ab4a1241967cfa5ffc

This code gives me a headache :exploding_head: (the old one, the new one - less)

**Review by sedited — APPROVED** [review]
ACK d4c7c4009da14c84576d43ab4a1241967cfa5ffc

**Comment by sr-gi** [comment]
> ACK [d4c7c40](https://github.com/bitcoin/bitcoin/commit/d4c7c4009da14c84576d43ab4a1241967cfa5ffc)
> 
> This code gives me a headache 🤯 (the old one, the new one - less)

This is literally the reason why I decided to fix this 🫠 

https://www.notion.so/srgi/PR-29415-110fd8ede41446fba41e81859bbdf985?pvs=4#521d59ebadd2475bb5630fbad147cbe4
